### PR TITLE
[ticket/9687] Refactor banning 

### PR DESCRIPTION
**Pull request by marc1706**
Continued work from #5393

Checklist:

- [x] Correct branch: master for new features; 3.3.x for fixes
- [x] Tests pass
- [x] Code follows coding guidelines: [master](https://area51.phpbb.com/docs/master/coding-guidelines.html) and [3.3.x](https://area51.phpbb.com/docs/dev/3.3.x/development/coding_guidelines.html)
- [x] Commit follows commit message [format](https://area51.phpbb.com/docs/dev/3.3.x/development/git.html)

Tracker ticket:

https://tracker.phpbb.com/browse/PHPBB3-9687

### CODE REVIEW

**Comment by rxu** [comment]
Probably use of this can help https://github.com/phpbb/phpbb/blob/282dd27e239e90c3b36e63dceb112fe933274677/tests/test_framework/phpbb_database_test_case.php#L380

File: tests/ban/ban_manager_test.php [hunk 1]
```diff
@@ -0,0 +1,686 @@
...
+			$this->expectException($expected_exception);
+		}
+
+		// Fix weird sorting by postgres
+		if (strtolower(substr(PHP_OS, 0, 3)) === 'linux' && is_array($expected) && $ban_type == 'email' && $this->db->get_sql_layer() == 'postgres')
+		{
+			$expected = array_reverse($expected);
+		}
+
+		$this->assertEquals($expected, $this->ban_manager->get_bans($ban_type));
```
Comment (rxu) [inline tests/ban/ban_manager_test.php:286]: Something like this fixes the test.

```suggestion
		$actual = $this->ban_manager->get_bans($ban_type);

		// Sort both arrays by ban_item to be synced
		if (is_array($expected) && !empty($actual))
		{
			usort($expected, function($a, $b)
				{
					return strcmp($a['ban_item'], $b['ban_item']) <=> 0;
				}
			);

			usort($actual, function($a, $b)
				{
					return strcmp($a['ban_item'], $b['ban_item']) <=> 0;
				}
			);
		}

		$this->assertEquals($expected, $actual);
```

Comment (marc1706) [reply tests/ban/ban_manager_test.php]: Thanks, that should solve it.

**Comment by marc1706** [comment]
Ready for review

File: phpBB/includes/functions_user.php [hunk 2]
```diff
@@ -968,482 +964,30 @@ function user_ban($mode, $ban, $ban_len, $ban_len_other, $ban_exclude, $ban_reas
...
-		$cache->destroy('sql', BANLIST_TABLE);
-
-		return true;
-	}
-
-	// There was nothing to ban/exclude. But destroying the cache because of the removal of stale bans.
-	$cache->destroy('sql', BANLIST_TABLE);
-
-	return false;
+	// TODO: logging
```
Comment (rxu) [inline phpBB/includes/functions_user.php:974]: Plan to implement within the current PR?

File: phpBB/phpbb/ban/type/base.php [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+		$this->sessions_table = $sessions_table;
+		$this->sessions_keys_table = $sessions_keys_table;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function set_user(\phpbb\user $user): void
+	{
+		// TODO: Implement new logging
```
Comment (rxu) [inline phpBB/phpbb/ban/type/base.php:64]: Plan to implement it later?

File: phpBB/phpbb/ban/type/email.php [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function prepare_for_storage(array $items): array
+	{
+		$this->get_excluded();
+
+		$regex = '#^.*?@.*|(([a-z0-9\-]+\.)+([a-z]{2,3}))$#i'; // TODO
```
Comment (rxu) [inline phpBB/phpbb/ban/type/email.php:43]: What is this TODO about?

**Review by rxu — COMMENTED** [review]
Also couldn't find unban algorithm in common even in the current code (or probably forgot where is it implemented) for temporary ban cases. Cron also doesn't seem to do it.

File: phpBB/includes/functions_user.php (hunk 2 above)
Comment (marc1706) [reply phpBB/includes/functions_user.php:974]: Done inside the ban manager, will remove this comment

File: phpBB/phpbb/ban/type/base.php (hunk 3 above)
Comment (marc1706) [reply phpBB/phpbb/ban/type/base.php]: It's included in the ban manager already, removing the todo

File: phpBB/phpbb/ban/type/email.php (hunk 4 above)
Comment (marc1706) [reply phpBB/phpbb/ban/type/email.php]: Not sure, although the regex was slightly different to before. Reverting back to the old one and removing the todo.

**Review by Derky — APPROVED** [review]
Thanks for the fixes, looks good!

**Pull request merged**